src/supernova/circuit.rs

```diff
@@ -0,0 +1,761 @@
+    &self,
+    mut cs: CS,
+    params: AllocatedNum<G::Base>,
+    i: AllocatedNum<G::Base>,
+    z_0: Vec<AllocatedNum<G::Base>>,
+    z_i: Vec<AllocatedNum<G::Base>>,
+    U: &[AllocatedRelaxedR1CSInstance<G>],
+    u: AllocatedR1CSInstance<G>,
+    T: AllocatedPoint<G>,
+    arity: usize,
+    last_augmented_circuit_index: &AllocatedNum<G::Base>,
+    program_counter: Option<AllocatedNum<G::Base>>,
+    num_augmented_circuits: usize,
+  ) -> Result<(AllocatedRelaxedR1CSInstance<G>, AllocatedBit), SynthesisError> {
+    // Check that u.x[0] = Hash(params, i, program_counter, U[], z0, zi)
+    let mut ro = G::ROCircuit::new(
+      self.ro_consts.clone(),
+      2 // params_next, i_new
+      + program_counter.as_ref().map(|_| 1).unwrap_or(0) // optional program counter
+        + 2 * arity // zo, z1
+        + num_augmented_circuits * (7 + 2 * self.params.n_limbs), // #num_augmented_circuits * (7 + [X0, X1]*#num_limb)
+    );
+    ro.absorb(&params);
+    ro.absorb(&i);
+    if let Some(program_counter) = program_counter.as_ref() {
+      ro.absorb(program_counter)
+    }
+
+    for e in &z_0 {
+      ro.absorb(e);
+    }
+    for e in &z_i {
+      ro.absorb(e);
+    }
+
+    U.iter().enumerate().try_for_each(|(i, U)| {
+      U.absorb_in_ro(cs.namespace(|| format!("absorb U {:?}", i)), &mut ro)
+    })?;
+
+    let hash_bits = ro.squeeze(cs.namespace(|| "Input hash"), NUM_HASH_BITS)?;
+    let hash = le_bits_to_num(cs.namespace(|| "bits to hash"), &hash_bits)?;
+    let check_pass: AllocatedBit = alloc_num_equals(
+      cs.namespace(|| "check consistency of u.X[0] with H(params, U, i, z0, zi)"),
+      &u.X0,
+      &hash,
+    )?;
+
+    // Run NIFS Verifier
+    let empty_U = AllocatedRelaxedR1CSInstance::alloc(
+      cs.namespace(|| "empty U"),
+      None,
+      self.params.limb_width,
+      self.params.n_limbs,
+    )?;
+    // select target when index match last_augmented_circuit_index, other left as empty
+    let U: Result<Vec<AllocatedRelaxedR1CSInstance<G>>, SynthesisError> = U
+      .iter()
+      .enumerate()
+      .map(|(i, U)| {
+        let i_alloc = alloc_const(
+          cs.namespace(|| format!("U_i i{:?} allocated", i)),
+          scalar_as_base::<G>(G::Scalar::from(i as u64)),
+        )?;
+        let equal_bit = Boolean::from(alloc_num_equals(
+          cs.namespace(|| format!("check U {:?} equal bit", i)),
+          &i_alloc,
+          last_augmented_circuit_index,
+        )?);
+        conditionally_select_alloc_relaxed_r1cs(
+          cs.namespace(|| format!("select on index namespace {:?}", i)),
+          U,
+          &empty_U,
+          &equal_bit,
+        )
+      })
+      .collect();
+
+    // Now reduce to one is safe, because only 1 of them != G::Zero based on the previous step
+    // TODO optimize this part to have raw linear-combination on variables to achieve less constraints
+    let U_to_fold = U?.iter().enumerate().try_fold(empty_U, |agg, (i, U)| {
```

Comment:
Hi, Thanks for catch up this and point out that!

Yes this is an under-constraints, due to there is no constraint for the `empty value` declaration.
I refered your idea and reviewed again of this logic.
Then I realized a better way to solve it is just rewrite this logic to avoid empty allocation, instead just iterate through the list to get the selected value when index match.

More detail in pseudo code
```
a := the array
index := target index
res = a[0] // default value is first element
res := index == 1 ? res:a[1]
res := index == 2 ? res:a[2] 
res := index == 3 ? res:a[3] 
...
return res
```
In particular, this ad-hoc N-reduce-1 optimisation is only applicable for this logic , giving assumption that prover have no choice but need to honestly choose correct witness `last_augmented_circuit_index` to hint the `fold(fresh instance, U[last_augmented_circuit_index])` in NIFS verify circuit, otherwise error will be catch up in the final CompressSNARK check.

Updated with https://github.com/microsoft/Nova/pull/204/commits/a938f293ddbf8dab782ed012f07cf8616965cd5b
Thanks again for the review :)

Reply:
Yes, I think this repeated conditional works. Given the size of the values (15 field elements for the allocated relax instance, I *think*), a multicase will be cheaper at about 6 clauses (i.e. 6 running instances). We don't need to worry about that now, but I think we can create an ergonomic and secure solution for this in general. I'll show my work later, perhaps elsewhere, so we can work through the details. For now, I think doing as you suggest would fairly simply solve the soundness issue.

Reply:
Issue follow-up: 

Updated a commit to fix relevant soundness issue https://github.com/microsoft/Nova/pull/204/commits/44758846c44c5d7950fa83554d2a453c8bf2129f

Background: there is existing and related soundness cause by `last_augmented_circuit_index` usage, once malicious prover set it outside range `[0, #num_opcode)`. More explicitly, It will cause no-matching of running instance selection in [the code]( https://github.com/microsoft/Nova/blob/f76dfc2d6276b50272feaeaae57e0f35dffe5838/src/supernova/circuit.rs#L447-L468), which means NO UPDATES of running instance*s* U[..]_i+1.

The solution I adopted is to propagate the processed `last_augmented_circuit_index` through whole circuit. The mindset behind is we already have ad-hoc logic to set  `last_augmented_circuit_index=0` once outside range. Then circuit need to use this new variable to assure `U[last_augmented_circuit_index]_i+1` updated accordingly.

Along with new fix, I also document more on this part https://github.com/microsoft/Nova/blob/44758846c44c5d7950fa83554d2a453c8bf2129f/src/supernova/circuit.rs#L411-L428 cc @srinathsetty if you can give more inputs :)